charms/kserve-controller/src/charm.py

```diff
@@ -362,105 +380,54 @@ def _generate_gateways_context(self) -> dict:
+        log.info("Generating certificates..")
+        certs = gen_certs(
+            service_name=self._controller_service_name,
+            namespace=self._namespace,
+            webhook_service=self._webhook_service_name,
         )
+        for k, v in certs.items():
+            setattr(self._stored, k, v)
+
+    def _upload_certs_to_container(
+        self, container: Container, destination_path: str, certs_store: StoredState
+    ) -> None:
+        """Upload generated certs to container.
+
+        Args:
+            container (Container): the container object to push certs to.
+            destination_path (str): path in str format where certificates will
+                be stored in the container.
+            certs_store (StoredState): an object where the certificate contents are stored.
+        """
+        try:
+            self._check_container_connection(container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
 
-        check_call(["openssl", "genrsa", "-out", "/run/ca.key", "2048"])
-        check_call(["openssl", "genrsa", "-out", "/run/server.key", "2048"])
-        check_call(
-            [
-                "openssl",
-                "req",
-                "-x509",
-                "-new",
-                "-sha256",
-                "-nodes",
-                "-days",
-                "3650",
-                "-key",
-                "/run/ca.key",
-                "-subj",
-                "/CN=127.0.0.1",
-                "-out",
-                "/run/ca.crt",
-            ]
-        )
-        check_call(
-            [
-                "openssl",
-                "req",
-                "-new",
-                "-sha256",
-                "-key",
-                "/run/server.key",
-                "-out",
-                "/run/server.csr",
-                "-config",
-                "/run/ssl.conf",
-            ]
-        )
-        check_call(
-            [
-                "openssl",
-                "x509",
-                "-req",
-                "-sha256",
-                "-in",
-                "/run/server.csr",
-                "-CA",
-                "/run/ca.crt",
-                "-CAkey",
-                "/run/ca.key",
-                "-CAcreateserial",
-                "-out",
-                "/run/cert.pem",
-                "-days",
-                "365",
-                "-extensions",
-                "v3_ext",
-                "-extfile",
-                "/run/ssl.conf",
-            ]
-        )
+        try:
+            container.push(f"{destination_path}/tls.key", certs_store.key, make_dirs=True)
+            container.push(f"{destination_path}/tls.crt", certs_store.cert, make_dirs=True)
+            container.push(f"{destination_path}/ca.crt", certs_store.ca, make_dirs=True)
+        except (ProtocolError, PathError) as e:
+            log.error(str(e))
+            self.unit.status = BlockedStatus(str(e))
```

Comment:
If we block here, it means we expect the user to resolve this somehow.  Is there something the user could do to fix this situation?

Reply:
I don't know what we can tell the user to do differently here to fix things, so in that case it feels like an Error rather than Blocked

Reply:
+1 for Error. This way next event might just resolve it when container becomes available.

Reply:
-1 for Error.
-1 for Blocked

we need to find a way to resolve it or defer the event. Why should we put unit into error
that is against charm best practice

Reply:
> -1 for Error. -1 for Blocked
> 
> we need to find a way to resolve it or defer the event. Why should we put unit into error that is against charm best practice

This is not event handling failing, but rather IO. Errors we are catching here are protocol and path related. If we defer event, I believe this wont be resolved.

Reply:
but that is the point. 
It is not blocked since user cannot do anything

but we should not intentionally pass Error. It is the same as have unexpected exception in python. In the ideal world charm should never go into error state

Reply:
According to the docs, the exceptions we are catching here are:
`ProtocolError` - Raised when there’s a higher-level protocol error talking to Pebble.
`PathError` - Raised when there’s an error with a specific path.
We cannot recover from either of those. Based on a discussion with @DnPlas, I suggest we raise a `GenericCharmRuntimeError` so the message is clear, and noted to @ca-scribner to make sure this kind of error handling is discussed as part of the base charm implementation.